src/node_click_handler.cpp

```diff
@@ -0,0 +1,190 @@
+//     * Neither the name of Southwest Research Institute® (SwRI®) nor the
+//       names of its contributors may be used to endorse or promote products
+//       derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL Southwest Research Institute® BE LIABLE 
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL 
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY 
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// *****************************************************************************
+
+#include <swri_console/node_click_handler.h>
+#include <swri_console/node_list_model.h>
+
+#include <ros/ros.h>
+#include <roscpp/GetLoggers.h>
+#include <roscpp/SetLoggerLevel.h>
+
+#include <QMessageBox>
+#include <QTextStream>
+
+namespace swri_console
+{
+  const std::string NodeClickHandler::ALL_LOGGERS = "All Loggers";
+  const std::string NodeClickHandler::GET_LOGGERS_SVC = "/get_loggers";
+  const std::string NodeClickHandler::SET_LOGGER_LEVEL_SVC = "/set_logger_level";
+
+  bool NodeClickHandler::eventFilter(QObject* obj, QEvent* event)
+  {
+    QContextMenuEvent* context_event;
+    QListView* list;
+
+    switch (event->type()) {
+      case QEvent::ContextMenu:
+        context_event = static_cast<QContextMenuEvent*>(event);
+        // First, make sure we clicked on the list and have an item in the list
+        // under the mouse cursor.
+        list = static_cast<QListView*>(obj);
+        if (list == NULL) {
+          return false;
+        }
+
+        return showContextMenu(list, context_event);
+      default:
+        // Pass through all other events
+        return QObject::eventFilter(obj, event);
+    }
+  }
+
+  bool NodeClickHandler::showContextMenu(QListView* list, QContextMenuEvent* event)
+  {
+    QModelIndexList index_list = list->selectionModel()->selectedIndexes();
+    if (index_list.isEmpty()) {
+      return false;
+    }
+
+    // Now get the node name that was clicked on and make a service call to
+    // get all of the loggers registered for that node.
+    NodeListModel* model = static_cast<NodeListModel*>(list->model());
+    node_name_ = model->nodeName(index_list.first());
+
+    std::string service_name = node_name_ + GET_LOGGERS_SVC;
+    ros::ServiceClient client = nh_.serviceClient<roscpp::GetLoggers>(service_name);
+    if (!client.waitForExistence(ros::Duration(2.0)))
```

Comment:
In practice, this call returns very quickly unless the roscore is hanging for some reason.  I put a timeout here only because I didn't want the GUI to hang forever if the roscore is stuck, but the duration is pretty arbitrary.  I'm more worried about making it *too* short because then it might be unusable over a slow network link.  Any suggestions for what would be more reasonable?